Highlight appears behind dictated text

mwyrzykowski opened the pull request:
#### 644d6535804c8d7a1ad1204e4203ba2462acf0a8
<pre>
Highlight appears behind dictated text
<a href="https://bugs.webkit.org/show_bug.cgi?id=257874">https://bugs.webkit.org/show_bug.cgi?id=257874</a>
radar://110492195

Reviewed by Aditya Keerthi.

264903@main introduced a regression where the following code:

  auto initialUnderlines = extractInitialUnderlines(string);
  underlines = !initialUnderlines.isEmpty() ? initialUnderlines : extractUnderlines(string);

was changed to:
  underlines = compositionUnderlines(string);

but compositionUnderlines doesn&apos;t populate the underlines when they are otherwise empty.

Re-introduce code removed from 264903@main and this issue appears resolved.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setAllowsInlinePredictions:]):
(-[WKWebViewConfiguration allowsInlinePredictions]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::extractUnderlines):
(WebKit::WebViewImpl::setMarkedText):

Canonical link: <a href="https://commits.webkit.org/265031@main">https://commits.webkit.org/265031@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b2d2ad681d1dc5e48490c4d4d592c0f6ec93c1f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9987 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11147 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9317 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11728 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9706 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12218 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9637 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10520 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8113 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11305 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7814 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16056 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8909 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8788 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12121 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9279 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8482 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/2300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12706 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9046 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -292,6 +292,16 @@ - (instancetype)init
     return self;
 }
 
+- (void)setAllowsInlinePredictions:(BOOL)enabled
```
264903@main (same change as the regression from the radar) caused this compiler warning. I just moved it out of the Private category to avoid the warning.

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4697,6 +4697,31 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
     return validAttributes.get().get();
 }
 
+static Vector<WebCore::CompositionUnderline> extractUnderlines(NSAttributedString *string)
```
This is the same as the stuff in the `#else` of  `compositionUnderlines`, right? If so, the `#else` section could probably be reduced to just a call to `extractUnderlines`

mwyrzykowski replied:
Good idea I didn't realize that. I'll just remove `extractUnderlines` and then this PR becomes a much smaller change.

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4697,6 +4697,31 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
 
+static Vector<WebCore::CompositionUnderline> extractUnderlines(NSAttributedString *string)
+{
+    Vector<WebCore::CompositionUnderline> result;
```
do we know what this function does that `compositionUnderlines` doesn't do that causes the broken behavior?

mwyrzykowski replied:
`compositionUnderlines` doesn't populate the result for dictated marked text, at least it didn't prior to the update in this PR.

rr-codes replied:
> compositionUnderlines doesn't populate the result for dictated marked text

Do we know why?

mwyrzykowski replied:
Yes, at some point it resulted in an extra blue underline under the dictated text. Not sure if that is still the case, but any change is outside the scope of an upstreaming PR.

mwyrzykowski replied:
Yes confirmed the underlines still incorrectly appear if I add this:
```
diff --git a/Source/WebKit/UIProcess/mac/WebViewImpl.mm b/Source/WebKit/UIProcess/mac/WebViewImpl.mm
index 94951f5b2259..5e453be761fb 100644
--- a/Source/WebKit/UIProcess/mac/WebViewImpl.mm
+++ b/Source/WebKit/UIProcess/mac/WebViewImpl.mm
@@ -5042,8 +5042,8 @@ static Vector<WebCore::CompositionUnderline> compositionUnderlines(NSAttributedS
 
         NSColor *underlineColor = attributes[NSUnderlineColorAttributeName];
         bool isClear = [underlineColor isEqual:NSColor.clearColor];
-
-        if (!isClear)
+        (void)isClear;
+        // if (!isClear)
```

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5062,7 +5062,11 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
     }
-#else
+
+    if (mergedUnderlines.size())
```
should probably be !mergedUnderlines.isEmpty()

mwyrzykowski replied:
Just wondering why would you prefer that when the implementation of isEmpty is `return !size()` ?

mwyrzykowski committed:
Highlight appears behind dictated text
https://bugs.webkit.org/show_bug.cgi?id=257874
radar://110492195

Reviewed by Aditya Keerthi.

264903@main introduced a regression where the following code:

  auto initialUnderlines = extractInitialUnderlines(string);
  underlines = !initialUnderlines.isEmpty() ? initialUnderlines : extractUnderlines(string);

was changed to:
  underlines = compositionUnderlines(string);

but compositionUnderlines doesn't populate the underlines when they are otherwise empty.

Re-introduce code removed from 264903@main and this issue appears resolved.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setAllowsInlinePredictions:]):
(-[WKWebViewConfiguration allowsInlinePredictions]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::extractUnderlines):
(WebKit::WebViewImpl::setMarkedText):

Canonical link: https://commits.webkit.org/265031@main
